internal/events/eventer.go

```diff
@@ -280,7 +290,55 @@ func (e *Eventer) Register(
 		topic,
 		e.webhookSubscriber,
 		func(msg *message.Message) error {
-			if err := handler(msg); err != nil {
+			if msg.Metadata == nil {
+				msg.Metadata = make(message.Metadata)
+			}
+
+			messageRetryCount := msg.Metadata.Get(MessageRetryCountKey)
+
+			// This check will most certainly never be true as currently the metadata is being added to every message,
+			// but it doesn't hurt to have it here as a failsafe.
+			// Same goes for the map initialisation above.
+			if messageRetryCount == "" {
+				messageRetryCount = "0"
+			}
+
+			messageRetryCountNumber, err := strconv.Atoi(messageRetryCount)
+			if err != nil {
+				e.router.Logger().Error("unable to convert messageRetryCount to int", err, watermill.LogFields{
+					"message_uuid": msg.UUID,
+					"topic":        topic,
+					"handler":      funcName,
+				})
+				return err
+			}
+
+			if messageRetryCountNumber >= maxMessageRetries {
+				e.router.Logger().Debug("maximum retries for message reached: adding message to DLQ", watermill.LogFields{
+					"message_uuid": msg.UUID,
+					"topic":        topic,
+					"handler":      funcName,
+					"max_retries":  maxMessageRetries,
+				})
+
+				err = e.webhookPublisher.Publish(DeadLetterQueueTopic, msg)
+				if err != nil {
+					e.router.Logger().Error("unable to publish message to dlq", err, watermill.LogFields{
+						"message_uuid": msg.UUID,
+						"topic":        topic,
+						"handler":      funcName,
+						"max_retries":  maxMessageRetries,
+					})
+					return err
+				}
+
+				return nil
+			}
+
+			messageRetryCountNumber++
+			msg.Metadata.Set(MessageRetryCountKey, fmt.Sprintf("%d", messageRetryCountNumber))
+
+			if err = handler(msg); err != nil {
```

Comment:
Rather than implementing this as 60 inline lines, what about the following:

1. Use [the `Retry` middleware](https://watermill.io/docs/middlewares/#retry) for retries.

   The documentation looks confusing, but I _think_ you want to use it like in https://github.com/stacklok/minder/pull/1842
3. Wrap the `Retry` middleware in a DLQ middleware.  I think [the `Poison` middleware](https://watermill.io/docs/middlewares/#poison) may be the Watermill term:

   ```golang
	pq, err := middleware.PoisonQueue(e, DeadLetterQueueTopic)
	
	router.AddMiddleware(
		pq,  // Run DLQ as the first / last handler, to handle all failed messages
		middleware.Retry{
			MaxRetries: 3,
			InitialInterval: time.Millisecond * 100,
			Logger: l,
		}.Middleware,
		middleware.CorrelationID,
   ```

Reply:
It would be nice if we could have a test for this -- create a router with a handler that always fails, and verify that the message ends up in the DLQ topic.

Reply:
The issue with the retry Middleware is that it doesn't work as advertised. It seems to ignore the retry limit and it just keeps retrying in a loop.

Reply:
Thanks, @evankanderson! The implementation is more straightforward now when utilising the PQ middleware. Also added a test for the DLQ scenario.